src/extension.ts

```diff
@@ -1763,29 +1770,81 @@ export async function saveUSSFile(doc: vscode.TextDocument, ussFileProvider: USS
     const remote = ending.substring(sesName.length).replace(/\\/g, "/");
 
     // get session from session name
-    let documentSession;
+    let documentSession: Session;
     let binary;
+    let node: ZoweUSSNode;
     const sesNode = (await ussFileProvider.mSessionNodes.find((child) => child.mProfileName && child.mProfileName.trim()=== sesName.trim()));
     if (sesNode) {
         documentSession = sesNode.getSession();
         binary = Object.keys(sesNode.binaryFiles).find((child) => child === remote) !== undefined;
     }
+    // Get specific node based on label and parent tree (session / favorites)
+    let nodes: ZoweUSSNode[];
+    if (!sesNode || sesNode.children.length === 0) {
+        // saving from favorites
+        nodes = utils.concatUSSChildNodes(ussFileProvider.mFavorites);
+    } else {
+        // saving from session
+        nodes = utils.concatUSSChildNodes([sesNode]);
+    }
+    node = await nodes.find((zNode) => {
+        if (zNode.contextValue === DS_FAV_TEXT_FILE_CONTEXT || zNode.contextValue === DS_TEXT_FILE_CONTEXT) {
+            return (zNode.fullPath.trim() === remote);
+        } else {
+            return false;
+        }
+    });
+
+    // define upload options
+    let etagToUpload: string;
+    let returnEtag: boolean;
+    if (node) {
+        etagToUpload = node.getEtag();
+        returnEtag = true;
+    }
 
     try {
-        const response = await vscode.window.withProgress({
+        const uploadResponse = await vscode.window.withProgress({
             location: vscode.ProgressLocation.Notification,
             title: localize("saveUSSFile.response.title", "Saving file...")
         }, () => {
-            return zowe.Upload.fileToUSSFile(documentSession, doc.fileName, remote, binary);  // TODO MISSED TESTING
+            return zowe.Upload.fileToUSSFile(documentSession, doc.fileName, remote, binary, null, etagToUpload, returnEtag);  // TODO MISSED TESTING
         });
-        if (response.success) {
-            vscode.window.showInformationMessage(response.commandResponse);
+        if (uploadResponse.success) {
+            vscode.window.showInformationMessage(uploadResponse.commandResponse);
+            // set local etag with the new etag from the updated file on mainframe
+            node.setEtag(uploadResponse.apiResponse.etag);
+        // this part never runs! zowe.Upload.fileToUSSFile doesn't return success: false, it just throws the error which is caught below!!!!!
         } else {
-            vscode.window.showErrorMessage(response.commandResponse);
+            vscode.window.showErrorMessage(uploadResponse.commandResponse);
         }
     } catch (err) {
-        log.error(localize("saveUSSFile.log.error.save", "Error encountered when saving USS file: ") + JSON.stringify(err));
-        vscode.window.showErrorMessage(err.message);
+        if (err.message.includes(localize("saveFile.error.ZosmfEtagMismatchError", "Rest API failure with HTTP(S) status 412"))) {
```

Comment:
Same as before. We may want to use some sort of message if or unique code that can be easily verified.
Again, no need to take action today